qualtran/resource_counting/symbolic_counting_utils.py

```diff
@@ -24,8 +25,45 @@
 document(SymbolicFloat, """A floating point value or a sympy expression.""")
 
 
+@frozen
+class Shaped:
+    """Symbolic value for an object that has a shape.
+
+    A Shaped object can be used as a symbolic replacement for any object that has an attribute `shape`,
+    for example numpy NDArrays.
+    Each dimension can be either an positive integer value or a sympy expression.
+
+    This is useful to do symbolic analysis of Bloqs whose call graph only depends on the shape of the input,
+    but not on the actual values.
+    For example, T-cost of the `QROM` Bloq depends only on the iteration length (shape) and not on actual data values.
+    """
+
+    shape: tuple[SymbolicInt, ...] = field(validator=validators.instance_of(tuple))
+
+    def is_symbolic(self):
+        return True
+
+
 def is_symbolic(*args) -> bool:
-    return any(isinstance(x, sympy.Basic) for x in args)
+    """Returns whether the inputs contain any symbolic object.
+
+    Returns:
+        True if any argument is either a sympy object,
+        or implements the `is_symbolic` method which returns True.
+    """
+
+    if len(args) != 1:
+        return any(is_symbolic(arg) for arg in args)
+
+    (arg,) = args
+    if isinstance(arg, sympy.Basic):
```

Comment:
I think we should be checking `isinstance(arg, sympy.Expr)`, right?

Reply:
I'm not sure actually, I just copied it from the existing code. 

It seems boolean expressions are instances of `sympy.Basic` but not `sympy.Expr` (https://docs.sympy.org/latest/explanation/glossary.html#term-Expr), which we may want to use for symbolic controls?

Reply:
Thanks. Interesting reading. Happy keeping this as Basic for now